L1Trigger/TrackFindingTMTT/src/TP.cc

```diff
@@ -0,0 +1,192 @@
+        phi0_(tpPtr->phi()),
+        vx_(tpPtr->vertex().x()),
+        vy_(tpPtr->vertex().y()),
+        vz_(tpPtr->vertex().z()),
+        d0_(vx_ * sin(phi0_) - vy_ * cos(phi0_)),                      // Copied from CMSSW class TrackBase::d0().
+        z0_(vz_ - (vx_ * cos(phi0_) + vy_ * sin(phi0_)) * tanLambda_)  // Copied from CMSSW class TrackBase::dz().
+  {
+    const vector<SimTrack>& vst = tpPtr->g4Tracks();
+    EncodedEventId eid = vst.at(0).eventId();
+    inTimeBx_ = (eid.bunchCrossing() == 0);  // TP from in-time or out-of-time Bx.
+    physicsCollision_ = (eid.event() == 0);  // TP from physics collision or from pileup.
+
+    this->fillUse();        // Fill use_ flag, indicating if TP is worth keeping.
+    this->fillUseForEff();  // Fill useForEff_ flag, indicating if TP is good for tracking efficiency measurement.
+  }
+
+  //=== Fill truth info with association from tracking particle to stubs.
+
+  void TP::fillTruth(const list<Stub>& vStubs) {
+    for (const Stub& s : vStubs) {
+      for (const TP* tp_i : s.assocTPs()) {
+        if (tp_i->index() == this->index())
+          assocStubs_.push_back(&s);
+      }
+    }
+
+    this->fillUseForAlgEff();  // Fill useForAlgEff_ flag.
+
+    this->calcNumLayers();  // Calculate number of tracker layers this TP has stubs in.
+  }
+
+  //=== Check if this tracking particle is worth keeping.
+  //=== (i.e. If there is the slightest chance of reconstructing it, so as to measure fake rate).
+
+  void TP::fillUse() {
+    constexpr bool useOnlyInTimeParticles = false;
+    constexpr bool useOnlyTPfromPhysicsCollisionFalse = false;
+    // Use looser cuts here those those used for tracking efficiency measurement.
+    // Keep only those TP that have a chance (allowing for finite track resolution) of being reconstructed as L1 tracks. L1 tracks not matching these TP will be defined as fake.
+
+    // Include all possible particle types here, as if some are left out, L1 tracks matching one of missing types will be declared fake.
+    static const vector<int> genPdgIdsAllUnsigned = {11, 13, 211, 321, 2212};
+    vector<int> genPdgIdsAll;
+    for (unsigned int i = 0; i < genPdgIdsAllUnsigned.size(); i++) {
+      genPdgIdsAll.push_back(genPdgIdsAllUnsigned[i]);
+      genPdgIdsAll.push_back(-genPdgIdsAllUnsigned[i]);
+    }
+
+    // Range big enough to include all TP needed to measure tracking efficiency
+    // and big enough to include any TP that might be reconstructed for fake rate measurement.
+    constexpr float ptMinScale = 0.7;
+    const float ptMin = min(settings_->genMinPt(), ptMinScale * settings_->houghMinPt());
+    constexpr double ptMax = 9.9e9;
+    const float etaExtra = 0.2;
+    const float etaMax = max(settings_->genMaxAbsEta(), etaExtra + std::abs(settings_->etaRegions()[0]));
+    constexpr double fixedVertRcut = 10.;
+    constexpr double fixedVertZcut = 35.;
+
+    static const TrackingParticleSelector trackingParticleSelector(ptMin,
```

Comment:
remaining static variable

Reply:
Since "static const" data is thread safe, and following the discussion leading to https://github.com/cms-sw/cmssw/pull/29381#issuecomment-628889782 , we propose to leave this code unchanged.